nx/lib/nx.ex

```diff
@@ -3546,11 +3597,316 @@ defmodule Nx do
   @doc type: :backend
   def backend_deallocate(tensor_or_container) do
     Nx.Defn.Composite.reduce(tensor_or_container, :ok, fn
-      %Nx.Tensor{} = tensor, :ok -> impl!(tensor).backend_deallocate(tensor)
-      _, :ok -> :ok
+      %Nx.Tensor{} = tensor, :ok ->
+        impl!(tensor).backend_deallocate(tensor)
+
+      _, :ok ->
+        :ok
     end)
   end
 
+  @doc """
+  Transforms the leading dimension in a tensor into a vectorized axis.
+
+  Each vectorization removes an axis from the shape and appends it to
+  vectorized axes list.
+
+  ## Examples
+
+      iex> t = Nx.iota({2, 3, 1, 1})
+      iex> vectorized = Nx.vectorize(t, :first)
+      #Nx.Tensor<
+        vectorized[first: 2]
+        s64[3][1][1]
+        [
+          [
+            [
+              [0]
+            ],
+            [
+              [1]
+            ],
+            [
+              [2]
+            ]
+          ],
+          [
+            [
+              [3]
+            ],
+            [
+              [4]
+            ],
+            [
+              [5]
+            ]
+          ]
+        ]
+      >
+      iex> Nx.vectorize(vectorized, :second)
+      #Nx.Tensor<
+        vectorized[first: 2][second: 3]
+        s64[1][1]
+        [
+          [
+            [
+              [0]
+            ],
+            [
+              [1]
+            ],
+            [
+              [2]
+            ]
+          ],
+          [
+            [
+              [3]
+            ],
+            [
+              [4]
+            ],
+            [
+              [5]
+            ]
+          ]
+        ]
+      >
+
+  A vectorized tensor can be thought of as a nested list of tensors
+  that have the shape equal to the tensor's base shape. Operations
+  are then applied in a nested manner accordingly.
```

Comment:
I am not sure if this explanation helps. :D  I would try something like:

    A vectorized axis applies each Nx operation to each entry in its dimension.